### PR TITLE
feat: Run subprocess with loading pattern

### PR DESCRIPTION
**Pull request by mildaniel**
#### Why is this change necessary?
When running long external subprocesses, sam-cli can appear to be "stuck" when in reality it's just waiting for the process.

#### How does it address the issue?
Runs the long subprocesses asynchronously and prints a timed pattern to the screen while waiting.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/hook_packages/terraform/hooks/prepare/hook.py [hunk 1]
```diff
@@ -112,20 +113,30 @@ def prepare(params: dict) -> dict:
...
+                # Terraform project expects input, and customer does not provide it.
+                command_args={
+                    "args": ["terraform", "plan", "-out", temp_file.name, "-input=false"],
+                    "cwd": terraform_application_dir,
+                    "stdout": PIPE,
+                }
             )
+            if isinstance(process_output, bytes):
+                process_output = process_output.decode("utf-8")
+            LOG.debug(process_output)
```
Comment (sriram-mv) [inline samcli/hook_packages/terraform/hooks/prepare/hook.py:138]: won't this LOG all the debug at once? can we stream it instead?

File: samcli/hook_packages/terraform/hooks/prepare/hook.py [hunk 2]
```diff
@@ -112,20 +113,30 @@ def prepare(params: dict) -> dict:
...
-
         with osutils.tempfile_platform_independent() as temp_file:
-            run(
-                # input false to avoid SAM CLI to stuck in case if the Terraform project expects input, and customer
-                # does not provide it.
-                ["terraform", "plan", "-out", temp_file.name, "-input=false"],
-                check=True,
-                capture_output=True,
-                cwd=terraform_application_dir,
+            process_output = invoke_subprocess_with_loading_pattern(
```
Comment (sriram-mv) [inline samcli/hook_packages/terraform/hooks/prepare/hook.py:127]: are we loading the `.` to stderr by default? There are CI runners that will error out that moment they see something printed to stderr, we have to be conscious of that.  

What is our current pattern for the dots, which stream do they land on ?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: samcli/hook_packages/terraform/hooks/prepare/hook.py (hunk 2 above)
Comment (mildaniel) [reply samcli/hook_packages/terraform/hooks/prepare/hook.py:127]: I was looking at other places in the codebase where we do something similar and we typically default to `stderr`. For example, we write loading dots when building Docker images https://github.com/aws/aws-sam-cli/blob/develop/samcli/local/docker/lambda_image.py#L236.  We can default it to `stdout` instead here if it makes more sense.

Comment (sriram-mv) [reply samcli/hook_packages/terraform/hooks/prepare/hook.py:127]: I see, maybe its less invasive that I thought it to be, still something to be conscious about.

Comment (moelasmar) [reply samcli/hook_packages/terraform/hooks/prepare/hook.py:127]: we use the stdout to write the `local invoke` output. I think to make it easier for people to read the lambda output programatically

File: samcli/hook_packages/terraform/hooks/prepare/hook.py (hunk 1 above)
Comment (moelasmar) [reply samcli/hook_packages/terraform/hooks/prepare/hook.py:138]: I think in this case (debug is enabled)  no idea of having the dots .

File: samcli/hook_packages/terraform/hooks/prepare/hook.py (hunk 2 above)
Comment (sriram-mv) [reply samcli/hook_packages/terraform/hooks/prepare/hook.py:127]: Fair, good callout.

File: samcli/hook_packages/terraform/hooks/prepare/hook.py (hunk 1 above)
Comment (mildaniel) [reply samcli/hook_packages/terraform/hooks/prepare/hook.py:138]: So you're saying we should stream one or the other? Is there a way of knowing what log level they have configured?

File: samcli/lib/utils/subprocess_utils.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        A function generating a pattern to the given stream
+    stream_writer: Optional[StreamWriter]
+        The stream to which to write the pattern
+
+    Returns
+    -------
+    str
+        A string containing the process output
+    """
+    stream_writer = stream_writer or StreamWriter(sys.stdout)
```
Comment (moelasmar) [inline samcli/lib/utils/subprocess_utils.py:66]: should it be `sys.stderr`, to leave the stdout for the lambda function output?

File: samcli/lib/utils/subprocess_utils.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            elif process.stdout:
+                # Logging level is DEBUG, streaming logs instead
+                for line in process.stdout:
+                    decoded_line = _check_and_process_bytes(line)
+                    LOG.debug(decoded_line)
+                    process_output += decoded_line
+
+            return_code = process.wait()
+
+            stream_writer.write("\n")
```
Comment (moelasmar) [inline samcli/lib/utils/subprocess_utils.py:84]: let's use `os.linesep` instead of `\n`

File: samcli/lib/utils/subprocess_utils.py (hunk 3 above)
Comment (mildaniel) [reply samcli/lib/utils/subprocess_utils.py:66]: Yup, good catch, didn't mean to commit that.

**Review by moelasmar — APPROVED** [review]

**Review by sriram-mv — APPROVED** [review]

**Pull request merged**